Percentage-based translations don't work with SVG <text>

nt1m opened the pull request:
#### b87209ee89e8e4595ec4f5409906b368310321cd
<pre>
Percentage-based translations don&apos;t work with SVG &lt;text&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=245122">https://bugs.webkit.org/show_bug.cgi?id=245122</a>
&lt;rdar://99854408&gt;

Reviewed by Myles C. Maxfield.

The two main issues are:
- Stale SVGTextElement::animatedLocalTransform() function:
Historically, all SVG elements have had the same animatedLocalTransform function, see: <a href="https://github.com/WebKit/WebKit/commit/f8cc29f9d2d96f5761a8aab9f6bd31afb0edf0a8">https://github.com/WebKit/WebKit/commit/f8cc29f9d2d96f5761a8aab9f6bd31afb0edf0a8</a>
This looks like an artifact of when SVGTextElement did not inherit from the common SVGGraphicsElement class, where we had to add the same function twice.
Removing this code puts it in-line with SVGGraphicsElement::animatedLocalTransform() which does take in account the reference box rect (for percentage based transforms) and transform-origin.

- We also need to compute transform at the end of RenderSVGText::layout, to avoid computing with the wrong bounds (which may happen if the font hasn&apos;t loaded yet).

Rebaseline tests resulting from these changes (mainly minor differences in dimensions that match other browsers).

Test: imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-002.html

* LayoutTests/TestExpectations:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/types-basicDOM-01-b-expected.txt:
* LayoutTests/platform/gtk/svg/batik/text/smallFonts-expected.txt:
* LayoutTests/platform/gtk/svg/batik/text/textFeatures-expected.txt:
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-2-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1-SE/types-dom-01-b-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/types-basicDOM-01-b-expected.txt:
* LayoutTests/platform/ios/svg/batik/text/textFeatures-expected.txt:
* LayoutTests/platform/ios/svg/text/append-text-node-to-tspan-expected.txt:
* LayoutTests/platform/ios/svg/text/modify-text-node-in-tspan-expected.txt:
* LayoutTests/platform/ios/svg/text/remove-text-node-from-tspan-expected.txt:
* LayoutTests/platform/ios/svg/text/scaled-font-expected.txt:
* LayoutTests/platform/ios/svg/text/scaling-font-with-geometric-precision-expected.txt:
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/types-basicDOM-01-b-expected.txt:
* LayoutTests/platform/mac/svg/batik/text/smallFonts-expected.txt:
* LayoutTests/platform/mac/svg/batik/text/textFeatures-expected.txt:
* LayoutTests/platform/mac/svg/text/scaling-font-with-geometric-precision-expected.txt:
* Source/WebCore/svg/SVGTextElement.cpp:
(WebCore::SVGTextElement::animatedLocalTransform const): Deleted.
* Source/WebCore/svg/SVGTextElement.h:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::layout):

Canonical link: <a href="https://commits.webkit.org/254656@main">https://commits.webkit.org/254656@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/977e690a471a50745936d640a6506d7c5de954f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89780 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34331 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99113 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155930 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93788 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32825 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93463 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95427 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76616 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/82137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/80963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30575 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30324 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3267 "Found 1 new test failure: security/mutation-observer-frame-detach.html (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33775 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/38785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1392 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32486 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -415,8 +408,15 @@ void RenderSVGText::layout()
-        updateCachedBoundariesInParents = oldBoundaries != objectBoundingBox();
+    } else {
+        if (m_needsTransformUpdate) {
+            m_localTransform = textElement().animatedLocalTransform();
```
```
AffineTransform SVGTextElement::animatedLocalTransform() const
{
    AffineTransform matrix;
    auto* style = renderer() ? &renderer()->style() : nullptr;

    // if CSS property was set, use that, otherwise fallback to attribute (if set)
    if (style && style->hasTransform()) {
        TransformationMatrix t;
        // For now, the transform-origin is not taken into account
        // Also, any percentage values will not be taken into account
        style->applyTransform(t, FloatRect(0, 0, 0, 0), RenderStyle::individualTransformOperations);
        // Flatten any 3D transform
        matrix = t.toAffineTransform();
    } else
        matrix = transform().concatenate();

    const AffineTransform* transform = const_cast<SVGTextElement*>(this)->supplementalTransform();
    if (transform)
        return *transform * matrix;
    return matrix;
}
```

It looks like this function uses style, but style should already have been computed before `layout()` is called, so that part of this change should have no effect.

It also looks like `supplementalTransform()` is the only code in here that could read layout information, but it looks like `supplementalTransform()` is only set for SMIL motion animation. However, `imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-002.html` doesn't seem to be using such an element. So I'm a bit confused about how this works.

nt1m replied:
This change makes it that we use https://searchfox.org/wubkat/rev/0c40ba62b482511fe03646f1d4982efd727475dd/Source/WebCore/svg/SVGGraphicsElement.cpp#78-123 instead of `SVGTextElement::animatedLocalTransform()`.

`SVGGraphicsElement::animatedLocalTransform()` uses `renderer()->transformReferenceBoxRect()` which relies on layout information, so it is critical that we finish performing layout *before* applying transforms.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -324,13 +324,6 @@ void RenderSVGText::layout()
     // textElement().updateLengthContext();
 
     bool updateCachedBoundariesInParents = false;
-    if (!isLayerBasedSVGEngineEnabled()) {
```
~~LBSE no longer makes use of the "m_localTransform" here, so you can avoid the computation in the section where you move the code to if LBSE is on~~. (you actually did that, I overlooked it)
Note: in LBSE currently transformations on <text> are ignored, your patch most likely fixes that too :-)

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -324,13 +324,6 @@ void RenderSVGText::layout()
     bool updateCachedBoundariesInParents = false;
-    if (!isLayerBasedSVGEngineEnabled()) {
-        if (m_needsTransformUpdate) {
-            m_localTransform = textElement().animatedLocalTransform();
```
Hmm, I naively expected a side-effect here, as SVG text layout code does access the 'm_localTransform' during layout, when computing the on-screen font-size scaling factor (which cries for a less involved replacment, but that's another patch...).

if the 'computeNewScaledFont...' code runs during layout(), you're using the SVG <text> 'transform' information from the _previous_ layout run (when it was last updated at the end of the last RenderSVGText::layout() cycle). Also first layout() misses the transform from the \<text\>, if specified.

So \<text transform="scale(10)" font-size="8px"/\> should be affected by this. If not, try a dynamic change of setting transform to scale(10) after the document loaded / painted. Please check if that is affected, to be on the safe side.

I'm asking as LBSE currently has issues with <text> + transform (latter is ignored), did not check the cause yet - still busy with compositing related work, before tackling bug fixes / features in LBSE.

nikolaszimmermann replied:
Note: in the other resolved commented you stated "It also looks like supplementalTransform() is the only code in here that could read layout information" -- that is correct and it is important to check who can read layout information, but it's equally important to check if 'm_localTransform' is actually accessed between the old location and the new one (where you moved the code to, below in the layout() function)).

nt1m replied:
I ran the suggested commands locally, the only changes found were the ones caught by EWS, the remainder was mostly noise, or flaky screenshotting.

nikolaszimmermann commented on Source/WebCore/svg/SVGTextElement.cpp:
```diff
@@ -44,30 +44,6 @@ Ref<SVGTextElement> SVGTextElement::create(const QualifiedName& tagName, Documen
     return adoptRef(*new SVGTextElement(tagName, document));
 }
 
-// We override SVGGraphics::animatedLocalTransform() so that the transform-origin
```
Yay for killing that code path!

nikolaszimmermann reviewed (commented):
Thanks for checking this! Be careful though that if your change affects repainting, you are not going to see all regressions when not using pixel tests (unfortunately we have thousands of valuable tests for e.g. repainting (svg/dynamic-updates/...) that depend on pixel testing, won't show a difference in text dumps -- at some point we have to convert all to reftests, but it's a huge task).

Anyhow: try to generate yourself a local pixel test baseline, before your patch and afterwards (run-webkit-tests --new-baseline --pixel-tests --tolerance=0 svg/  + apply your patch, rebuild + run run-webkit-tests --pixel-tests svg/). Then you can really tell the impact of your patch - please give it a try and report.

litherum reviewed (approved):
r=me assuming `transform-origin` on SVG `<text>` is well-tested.

nt1m committed:
Percentage-based translations don't work with SVG <text>
https://bugs.webkit.org/show_bug.cgi?id=245122
<rdar://99854408>

Reviewed by Myles C. Maxfield.

The two main issues are:
- Stale SVGTextElement::animatedLocalTransform() function:
Historically, all SVG elements have had the same animatedLocalTransform function, see: https://github.com/WebKit/WebKit/commit/f8cc29f9d2d96f5761a8aab9f6bd31afb0edf0a8
This looks like an artifact of when SVGTextElement did not inherit from the common SVGGraphicsElement class, where we had to add the same function twice.
Removing this code puts it in-line with SVGGraphicsElement::animatedLocalTransform() which does take in account the reference box rect (for percentage based transforms) and transform-origin.

- We also need to compute transform at the end of RenderSVGText::layout, to avoid computing with the wrong bounds (which may happen if the font hasn't loaded yet).

Rebaseline tests resulting from these changes (mainly minor differences in dimensions that match other browsers).

Test: imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-002.html

* LayoutTests/TestExpectations:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/types-basicDOM-01-b-expected.txt:
* LayoutTests/platform/gtk/svg/batik/text/smallFonts-expected.txt:
* LayoutTests/platform/gtk/svg/batik/text/textFeatures-expected.txt:
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-2-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1-SE/types-dom-01-b-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/types-basicDOM-01-b-expected.txt:
* LayoutTests/platform/ios/svg/batik/text/textFeatures-expected.txt:
* LayoutTests/platform/ios/svg/text/append-text-node-to-tspan-expected.txt:
* LayoutTests/platform/ios/svg/text/modify-text-node-in-tspan-expected.txt:
* LayoutTests/platform/ios/svg/text/remove-text-node-from-tspan-expected.txt:
* LayoutTests/platform/ios/svg/text/scaled-font-expected.txt:
* LayoutTests/platform/ios/svg/text/scaling-font-with-geometric-precision-expected.txt:
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/types-basicDOM-01-b-expected.txt:
* LayoutTests/platform/mac/svg/batik/text/smallFonts-expected.txt:
* LayoutTests/platform/mac/svg/batik/text/textFeatures-expected.txt:
* LayoutTests/platform/mac/svg/text/scaling-font-with-geometric-precision-expected.txt:
* Source/WebCore/svg/SVGTextElement.cpp:
(WebCore::SVGTextElement::animatedLocalTransform const): Deleted.
* Source/WebCore/svg/SVGTextElement.h:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::layout):

Canonical link: https://commits.webkit.org/254656@main
